x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,218 @@
+    private final Object ioLock = new Object();
+    private volatile IOControl savedIoControl;
+
+    StreamingHttpResultPublisher(ThreadPool threadPool, HttpSettings settings, ActionListener<Flow.Publisher<HttpResult>> listener) {
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.settings = Objects.requireNonNull(settings);
+        this.listener = Objects.requireNonNull(listener);
+    }
+
+    @Override
+    public void responseReceived(HttpResponse httpResponse) throws IOException {
+        this.response = httpResponse;
+        this.resultQueue.offer(HttpResult.create(settings.getMaxResponseSize(), response));
+        this.listener.onResponse(this);
+    }
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super HttpResult> subscriber) {
+        this.subscriber = subscriber;
+        subscriber.onSubscribe(new HttpSubscription());
+    }
+
+    @Override
+    public void consumeContent(ContentDecoder contentDecoder, IOControl ioControl) throws IOException {
+        if (isDone() == false) { // if the subscriber canceled us, don't bother to keep reading
+            var buffer = new SimpleInputBuffer(4096);
+            var consumed = buffer.consumeContent(contentDecoder);
+            var allBytes = new byte[consumed];
+            buffer.read(allBytes);
+
+            // we can have empty bytes, don't bother sending them
+            if (allBytes.length > 0) {
+                resultQueue.offer(new HttpResult(response, allBytes));
+            }
+
+            // always check if totalByteSize > the configured setting in case the settings change
+            if (totalByteSize.accumulateAndGet(allBytes.length, Long::sum) >= settings.getMaxResponseSize().getBytes()) {
+                pauseProducer(ioControl);
+            }
+
+            // always run in case we're waking up from a pause and need to start a new thread
+            runOrScheduleNextMessage();
+        }
+    }
+
+    private void pauseProducer(IOControl ioControl) {
+        ioControl.suspendInput();
+        synchronized (ioLock) {
+            savedIoControl = ioControl;
+        }
+    }
+
+    private void resumeProducer() {
+        synchronized (ioLock) {
+            if (savedIoControl != null) {
+                savedIoControl.requestInput();
+                savedIoControl = null;
+            }
+        }
+    }
+
+    // if there is currently a thread running in a loop, it should pick up this new request
+    // if not, check if this thread is one of ours and reuse it
+    // else, offload to a new thread so we do not block Apache's connection thread or ElasticSearch's transport thread
+    private void runOrScheduleNextMessage() {
+        if (subscriber != null && isRunning.compareAndSet(false, true)) {
+            var currentThreadPool = EsExecutors.executorName(Thread.currentThread().getName());
+            if (UTILITY_THREAD_POOL_NAME.equalsIgnoreCase(currentThreadPool)) {
+                new OffloadThread().run();
```

Comment:
Is it an error if more than 1 thread executes this code at the same time? It would mean that 2 threads are calling this code at the same time.

```
               var nextRequest = resultQueue.poll();
                subscriber.onNext(nextRequest);
```

Is `subscriber.onNext()` thread safe? If it is then it may be easier to always schedule a new thread and avoid the tricky `isRunning` checks. There could be a race where the loop in line 200 has finished but `isRunning.set(false);` on line 205 is not set yet

Reply:
I feel like there's an expectation that subscribers are thread safe? My thinking is that it's in the `java.util.concurrent` namespace so it *should* be thread safe?

In all the implementations of Flow.* that I've seen, they all use some version of a state machine to control when the subscriber is called to effectively implement that thread safety.  `isRunning` was how I thought to do that as well.

> There could be a race where the loop in line 200 has finished but isRunning.set(false); on line 205 is not set yet

This sounds right, there's an expectation that `onSubscribe` and `onNext` will eventually call `Subscription.request` which will increase demand, but if we haven't done that yet then there may be supply (apache) toggling the `isRunning` with no demand and the first demand "misses" scheduling.  Now I'm wondering if there's a bug in the JDK's HTTP client since they use a similar mechanism, but I mean how often are we at the end of the supply queue without ever having demand?

Let me rework this a bit to get rid of `isRunning` and do something like "is there both supply and demand"

Reply:
> Now I'm wondering if there's a bug in the JDK's HTTP client since they use a similar mechanism, but I mean how often are we at the end of the supply queue without ever having demand?

Ah, JDK wraps it in a `while(true)` loop until supply can schedule the thread